nomad/structs/node_pool.go

```diff
@@ -0,0 +1,131 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package structs
+
+import (
+	"fmt"
+	"regexp"
+
+	"github.com/hashicorp/go-multierror"
+)
+
+const (
+	// NodePoolAll is a reserved node pool that always includes all nodes in
+	// the cluster.
+	NodePoolAll            = "all"
+	NodePoolAllDescription = "Node pool with all nodes in the cluster."
+
+	// NodePoolDefault is a reserved node pool for nodes that don't specify a
+	// node pool in their configuration.
+	NodePoolDefault            = "default"
+	NodePoolDefaultDescription = "Default node pool."
+
+	// maxNodePoolDescriptionLength is the maximum length allowed for a node
+	// pool description.
+	maxNodePoolDescriptionLength = 256
+)
+
+var (
+	// validNodePoolName is the rule used to validate a node pool name.
+	validNodePoolName = regexp.MustCompile("^[a-zA-Z0-9]{1,128}$")
+)
+
+// NodePool allows partioning infrastructure
+type NodePool struct {
+	// Name is the node pool name. It must be unique.
+	Name string
+
+	// Description is the human-friendly description of the node pool.
+	Description string
+
+	// Meta is a set of user-provided metadata for the node pool.
+	Meta map[string]string
```

Comment:
Oh interesting, the other `Canonicalize` methods actually set them to `nil`:
https://github.com/hashicorp/nomad/blob/17633d83704c73174cde0a75683a37d36fcd192f/nomad/structs/structs.go#L4521-L4525

The reason seems be for consistent diffs of `nil` vs. empty, so I don't think it would mater much if we initialize an empty map. Maybe it adds more storage space, but for node pools that's probably negligible as I don't expect there to be that many, and a `nil` map is a recipe for panics 😅